### PR TITLE
create a separate make_websoket_handler method

### PR DESCRIPTION
**Pull request by mortbauer**
This allows the following pattern:
```
@app.route('/ws/devices/<device_id>')
@auth_required
@app.make_websocket_handler
async def websocket_connection(request,ws,device_id):
```
Which means that the authentication can be checked before the websocket upgrade is started, and thus also responding with a proper 401  Unauthorized.

The current behaviour is not altered by this change.

### CODE REVIEW

File: sanic/app.py [hunk 1]
```diff
@@ -457,6 +457,49 @@ def add_route(
...
+
+                pass
+
+            if self.asgi:
+                ws = request.transport.get_websocket_connection()
+            else:
+                try:
+                    protocol = request.transport.get_protocol()
+                except AttributeError:
+                    # On Python3.5 the Transport classes in asyncio do not
```
Comment (ahopkins) [inline sanic/app.py]: Sanic does not aim to support Python 3.5

**Comment by sjsadowski** [comment]
A few minor linting issues to clean up and please add a test around this given that we're restructuring websockets here with a separate function.

File: sanic/app.py [hunk 2]
```diff
@@ -457,6 +457,49 @@ def add_route(
         )(handler)
         return handler
 
+    def make_websocket_handler(self,handler,subprotocols=None):
```
Comment (ahopkins) [inline sanic/app.py]: I want to think on this name a little. We might be able to find something more approachable and understandable.

**Comment by ahopkins** [comment]
I put some more thought in this...

Rather than add a new handler, I think a better solution is either to allow this on the `@app.websocket` handler (probably more difficult) or add a `@app.route(websocket=True)` argument.

**Comment by mortbauer** [comment]
The problem with doing that via the `@app.websocket` is that `@app.websocket` does expect the uri as first argument, which would be not used if one just wants to turn the request handler into a websocket handler -> possible but in my opinion not good

The problem with integrating it with a `@app.route(websocket=True)` is the same as the original one since it would wrap and add the route in one function call, however what I want is to split this functionality.

1. wrap to produce a websocket handler
... wrap with other decorators
2. add a route for the handler

File: sanic/app.py (hunk 2 above)
Comment (mortbauer) [reply sanic/app.py]: what about websockify ?

**Comment by ahopkins** [comment]
I think this will need to be a decision with some more discussion and weight from @huge-success/sanic-core-devs since it is a bit of an API change. 

Thoughts from anyone else?

**Comment by sjsadowski** [comment]
I'm ultimately fine with it so long as the previous method is implemented to be compatible. I'm also fine with making a breaking change so long as it is documented. If we want to discuss in depth we should start a thread on the forum.

**Pull request closed without merging**